docs/protocol/core/v3.0.rst

```diff
@@ -103,11 +103,72 @@ Data types
 
 TODO define core data types
 
-Regular chunk grids
--------------------
-
-TODO define regular chunk grids, including how to form a key for each chunk in a grid
-
+Chunk grids
+-----------
+
+A chunk grid defines a set of chunks which contain the elements of an
+array. The chunks of a grid form a tessellation of the array space,
+which is a space defined by the dimensionality and shape of the
+array. This means that every element of the array is a member of one
+chunk, and there are no gaps or overlaps between chunks.
+
+In general there are different possible types of grids. The core
+protocol defines the regular grid type, where all chunks are
+hyperrectangles of the same shape. Protocol extensions may define
+other grid types, such as rectilinear grids where chunks are still
+hyperrectangles but do not all share the same shape.
+
+A grid type also defines rules for constructing a unique key for each
+chunk, which is a string of ASCII characters that can be used to save
+and retrieve chunk data in a store.
+
+Regular grids
+~~~~~~~~~~~~~
+
+A regular grid is a type of grid where an array is divided into chunks
+such that each chunk is a hyperrectangle of the same shape. The
+dimensionality of the grid is the same as the dimensionality of the
+array. Each chunk in the grid can be addressed by a tuple of positive
+integers (i, j, k, ...) corresponding to the indices of the chunk
+along each dimension.
+
+The origin vertex of a chunk has coordinates in the array space (i *
+dx, j * dy, k * dz, ...) where (dx, dy, dz, ...) are the grid spacings
+along each dimension, also known as the chunk shape. Thus the origin
+vertex of the chunk at grid index (0, 0, 0, ...) is at coordinate (0,
+0, 0, ...) in the array space, i.e., the grid is aligned with the
+origin of the array. If the length of any array dimension is not
+perfectly divisible by the chunk length along the same dimension, then
+the grid will overhang the edge of the array space.
+
+The shape of the chunk grid will be (ceil(x / dx), ceil(y / dy),
+ceil(z / dz), ...)  where (x, y, z, ...) is the array shape, / is the
+division operator and ceil() is the ceiling function. For example, if
+a 3 dimensional array has shape (10, 200, 3000), and has chunk shape
+(5, 20, 400), then the shape of the chunk grid will be (2, 10, 8),
+meaning that there will be 2 chunks along the first dimension, 10
+along the second dimension, and 8 along the third dimension.
+
+An element of an array with coordinates (i, j, k, ...) will occur
+within the chunk at grid index (i // dx, j // dy, k // dz, ...), where
+// is the floor division operator. The element will have coordinates
+(i % dx, j % dy, k % dz, ...) within that chunk. For example, @@TODO
+example.
+
+The key for chunk with grid index (i, j, k, ...) is formed by
+concatenating the ASCII string representation of each index, joined
+together via the period (".") character. For example, in a 3
```

Comment:
Related to the "tuple of strings" from above, I would propose either leaving the joining operation to the backend or going so far as to suggest "/" as the default. My understanding is that cloud storage doesn't suffer under use of "/" but local storage _does_ suffer under use of ".".

Reply:
I've just tentatively pushed an edit which allows for an array to have configurable prefix, suffix and separator for chunk keys, which would be a mechanism to allow e.g. use of "/" as chunk key separator. Happy to row back on this if anyone feels that should be a protocol extension.